twine/package.py

```diff
@@ -75,7 +88,7 @@ def __init__(
         self.metadata = metadata
         self.python_version = python_version
         self.filetype = filetype
-        self.safe_name = _safe_name(metadata.name)
+        self.safe_name = _safe_name(cast(str, metadata.name))
```

Comment:
I don't know this part of the codebase super well, but: if `metadata.name` can be `None`, shouldn't we fail with `InvalidDistribution` here rather than casting and percolating the invalid state through `_safe_name`?

Reply:
It occurs to me that maybe better than a cast would be to write a protocol that captures this expectation explicitly. I'll explore some options.

Reply:
I tried [making a protocol](https://gist.github.com/jaraco/22c22564a9fe6fdb137ef0a217511030) to specifically refine the `pkginfo.Distribution` interface, but it fails because:

```
twine/package.py:77: error: All bases of a protocol must be protocols  [misc]
```

However, if I simply remove the protocol, it seems to work. Fixed in 4cf7f84.